### PR TITLE
Enable calling with metadata

### PR DESCRIPTION
**Pull request by noodlesblue**
Adding `callWithMetadata` that allows accepts metadata. This also includes the fix for a bug handling semicolons in the proto file.

### CODE REVIEW

**Comment by daroczig** [comment]
the unit test seems to be broken with the 3.x.x version of `processx` -- seems to be unrelated to this PR

File: R/client.R [hunk 1]
```diff
@@ -15,6 +15,7 @@ grpc_client <- function(impl, channel) {
       
       list(
         call = function(x) read(ResponseDescriptor, fetch(channel, fn$name, serialize(x, NULL))),
+        callWithMetadata = function(x, metadata) read(ResponseDescriptor, fetchWithMetadata(channel, fn$name, serialize(x, NULL), metadata)),
```
Comment (nfultz) [inline R/client.R]: Instead of two functions, can we just do
```
call = function(x, metadata=character(0)) 
```

File: R/parser.R [hunk 2]
```diff
@@ -35,7 +35,7 @@ read_services <- function(file){
 
     w <- "RequestType"
 
-    while(tokens[i] != '}'){
+    while(tokens[i] != '}' && tokens[i] != ';'){
```
Comment (nfultz) [inline R/parser.R]: We / I should add some more tests around parsing.

File: src/client.cpp [hunk 3]
```diff
@@ -65,6 +65,19 @@ RawVector fetch(CharacterVector server, CharacterVector method, RawVector reques
   // perform request
   grpc_op ops[6];
   grpc_op *op;
+
+  int METADATA_LENGTH = metadata.length() / 2;
+  grpc_metadata meta_c[METADATA_LENGTH];
+
+  if (METADATA_LENGTH > 0) {
```
Comment (nfultz) [inline src/client.cpp]: I think you don't need the if statement here, flow won't enter the loop when length is 0 anyway.

File: src/client.cpp [hunk 4]
```diff
@@ -65,6 +65,19 @@ RawVector fetch(CharacterVector server, CharacterVector method, RawVector reques
   // perform request
   grpc_op ops[6];
   grpc_op *op;
+
+  int METADATA_LENGTH = metadata.length() / 2;
```
Comment (nfultz) [inline src/client.cpp]: We can check for null here also.

No need to capslock either.

File: src/client.cpp [hunk 5]
```diff
@@ -91,8 +104,16 @@ RawVector fetch(CharacterVector server, CharacterVector method, RawVector reques
   memset(ops, 0, sizeof(ops));
   op = ops;
   op->op = GRPC_OP_SEND_INITIAL_METADATA;
-  op->data.send_initial_metadata.count = 0;
+  if (METADATA_LENGTH > 0) {
+    op->data.send_initial_metadata.count = METADATA_LENGTH;
```
Comment (nfultz) [inline src/client.cpp]: ```op->data.send_initial_metadata.count = METADATA_LENGTH```
can be pulled outside of the if/else.

File: src/client.cpp [hunk 6]
```diff
@@ -28,7 +28,7 @@ RawVector sliceToRaw2(grpc_slice slice){
 
 
 // [[Rcpp::export]]
-RawVector fetch(CharacterVector server, CharacterVector method, RawVector requestArg) {
+RawVector fetchWithMetadata(CharacterVector server, CharacterVector method, RawVector requestArg, CharacterVector metadata) {
```
Comment (nfultz) [inline src/client.cpp]: Let's call this fetch, and get rid of the wrapper function at the bottom.

File: demo/helloclient.R [hunk 7]
```diff
@@ -12,10 +12,21 @@ client <- grpc_client(impl, "localhost:50051")
 
 
 for(who in c("Neal", "Gergely")){
-    who <- client$SayHello$build(name=who)
+    hello <- client$SayHello$build(name=who)
```
Comment (nfultz) [inline demo/helloclient.R]: Feel free to add yourself to `who` :)

**Comment by nfultz** [comment]
@sieverssj @noodlesblue Thank you for the patch - overall I like it quite a bit. I think it would be simpler to make metadata an optional argument on `fetch` rather than having separate `fetch` and `fetchWithMetadata` functions. All my other comments are minor/style preferences.

File: R/parser.R (hunk 2 above)
Comment (noodlesblue) [reply R/parser.R]: Sounds good.

File: src/client.cpp (hunk 4 above)
Comment (noodlesblue) [reply src/client.cpp]: I'll remove the capslock, but I can't for the life of me seem to figure out how to check for null...

File: demo/helloclient.R (hunk 7 above)
Comment (noodlesblue) [reply demo/helloclient.R]: Cool! Thanks!

**Comment by nfultz** [comment]
Awesome! I'm going to merge this in and bump the version.

**Pull request merged**